controllers/workspace/provision/automount.go

```diff
@@ -0,0 +1,237 @@
+				DefaultMode: &modeReadOnly,
+			},
+		},
+	}
+	return workspaceVolumeMount
+}
+
+func getAutoMountConfigMapVolumeMount(mountPath, name string) corev1.VolumeMount {
+	workspaceVolumeMount := corev1.VolumeMount{
+		Name:      common.AutoMountConfigMapVolumeName(name),
+		ReadOnly:  true,
+		MountPath: mountPath,
+	}
+	return workspaceVolumeMount
+}
+
+func getAutoMountSecretVolumeMount(mountPath, name string) corev1.VolumeMount {
+	workspaceVolumeMount := corev1.VolumeMount{
+		Name:      common.AutoMountSecretVolumeName(name),
+		ReadOnly:  true,
+		MountPath: mountPath,
+	}
+	return workspaceVolumeMount
+}
+
+func getAutoMountConfigMapEnvFromSource(name string) corev1.EnvFromSource {
+	return corev1.EnvFromSource{
+		ConfigMapRef: &corev1.ConfigMapEnvSource{
+			LocalObjectReference: corev1.LocalObjectReference{
+				Name: name,
+			},
+		},
+	}
+}
+
+func getAutoMountSecretEnvFromSource(name string) corev1.EnvFromSource {
+	return corev1.EnvFromSource{
+		SecretRef: &corev1.SecretEnvSource{
+			LocalObjectReference: corev1.LocalObjectReference{
+				Name: name,
+			},
+		},
+	}
+}
+
+func checkAutoMountVolumesForCollision(base, automount []v1alpha1.PodAdditions) error {
+	// Get a map of automounted volume names to volume structs
+	automountVolumeNames := map[string]corev1.Volume{}
+	for _, podAddition := range automount {
+		for _, volume := range podAddition.Volumes {
+			automountVolumeNames[volume.Name] = volume
+		}
+	}
+
+	// Check that workspace volumes do not conflict with automounted volumes
+	for _, podAddition := range base {
+		for _, volume := range podAddition.Volumes {
+			if conflict, exists := automountVolumeNames[volume.Name]; exists {
+				return fmt.Errorf("DevWorkspace volume %s conflicts with automounted volume from %s",
+					volume.Name, formatVolumeDescription(conflict))
+			}
+		}
+	}
+
+	// Check that automounted mountPaths do not collide
+	automountVolumeMountsByMountPath := map[string]corev1.VolumeMount{}
+	for _, podAddition := range automount {
+		for _, vm := range podAddition.VolumeMounts {
+			if conflict, exists := automountVolumeMountsByMountPath[vm.MountPath]; exists {
+				return fmt.Errorf("auto-mounted volumes from %s and %s have the same mount path",
+					getVolumeDescriptionFromVolumeMount(vm, automount), getVolumeDescriptionFromVolumeMount(conflict, automount))
+			}
+			automountVolumeMountsByMountPath[vm.MountPath] = vm
+		}
+	}
+
+	// Check that automounted volume mountPaths do not conflict with existing mountPaths in any container
+	for _, podAddition := range base {
+		for _, container := range podAddition.Containers {
+			for _, vm := range container.VolumeMounts {
+				if conflict, exists := automountVolumeMountsByMountPath[vm.MountPath]; exists {
+					return fmt.Errorf("DevWorkspace volume %s in container %s has same mountpath as auto-mounted volume from %s",
+						getVolumeDescriptionFromVolumeMount(vm, base), container.Name, getVolumeDescriptionFromVolumeMount(conflict, automount))
```

Comment:
Do you have a check on the previous stage that different volumes are not mounted on the same path? Probably not or at least I haven't found it easily.
If we don't - please create an issue to validate flattened devfile on any kind of conflicts

Reply:
Good point -- we do some checking but it's not consistent. Created https://github.com/devfile/devworkspace-operator/issues/497